google-cloud-spanner/lib/google/cloud/spanner/transaction.rb

```diff
@@ -1111,12 +1137,63 @@ def self.from_grpc grpc, session
           end
         end
 
+        ##
+        # @private Checks if a transaction is already created.
+        def existing_transaction?
+          !no_existing_transaction?
+        end
+
+        ##
+        # @private Checks if transaction is not already created.
+        def no_existing_transaction?
+          @grpc.nil?
+        end
+
         protected
 
+        ##
+        # @private Facilitates a thread-safe execution of an rpc
+        # for inline-begin of a transaction. This method is optimised to
+        # use mutexes only when necessary, while still acheiving thread-safety.
+        #
+        # Note: Do not use @seqno directly while using this method. Instead, use
+        # the seqno variable passed to the block.
+        def safe_execute
+          loop do
+            if existing_transaction?
+              # Create a local copy of @seqno to avoid concurrent
+              # operations overriding the incremented value.
+              seqno = next_seqno
+              # If a transaction already exists, execute rpc without mutex
+              return yield seqno
+            end
+
+            @mutex.synchronize do
+              next if existing_transaction?
+              @seqno += 1
+              return yield @seqno
+            end
+          end
+        end
+
+        ##
+        # Create a new transaction in a thread-safe manner,
+        def safe_begin_transaction
+          @mutex.synchronize do
+            return if existing_transaction?
+            ensure_session!
+            @grpc = service.begin_transaction session.path
+          end
+        end
+
         # The TransactionSelector to be used for queries
```

Comment:
It would probably be good to note in the documentation here that calls to this method must be synchronized by the mutex, since it includes logic that depends on the state of `existing_transaction?` which could otherwise be changed by another thread that owns the mutex. (Currently, it is properly called from within synchronize blocks because all calls are from blocks passed to `safe_execute`. We might even want to mention that that is the expected usage.)

Reply:
I updated the documentation with the suggested info.